Add fetch_team_stats function and documentation

DanielTomaro13 opened the pull request:
Added a function to scrape team stats totals and averages from AFLtables

DanielTomaro13 committed:
Add fetch_team_stats function and documentation

DanielTomaro13 commented:
Hi there! 
First timer and relative newbie so please excuse me.
I have been working and loving Fitzroy for a little bit now and often calculate team stats manually. 
I am now looking to get them from different sources, firstly footywire. The website wheeloratings.com has some pretty cool team stats so would be good to be able to access them as they are quite detailed and advanced
Cheers

jimmyday12 commented on R/fetch_team_stats.R:
```diff
@@ -0,0 +1,100 @@
+# Declare globals to avoid R CMD check notes
+utils::globalVariables(c("Team", "type", "Home.Team", "Away.Team", "Games", "where", "all_of", "setNames"))
```
@DanielTomaro13 rather than delcaring these like this, I've been using `.data$variable` when referencing these things as it's a bit more 

You can see an example in the file `fetch_fixture.R` on line 185

```
  # Remove those columns
  df <- df |>
    dplyr::filter(!.data$IsRound) |>
    dplyr::filter(!.data$IsHeader) |>
    dplyr::select(-"IsRound", -"IsHeader") |>
    dplyr::filter(.data$X1 != "")
``` 

See https://dplyr.tidyverse.org/articles/in-packages.html

jimmyday12 commented on R/fetch_team_stats.R:
```diff
@@ -0,0 +1,100 @@
+#' @examples
+#' fetch_team_stats(2024)
+#' fetch_team_stats(2023, summary_type = "averages")
+fetch_team_stats <- function(season, summary_type = "totals") {
```
@DanielTomaro13 If you could re-write this using the format we've got for other functions that would be great. 

The general idea is that 

`fetch_team_stats` should be a general function that could use multiple sources. 

The function you have written here should be renamed `fetch_team_stats_afltables` and then `fetch_team_stats` should call this function when the `source = "afltables"`. 

You should be able to copy paste from any function, but the example from `fetch_player_stats` is below

```
fetch_player_stats <- function(season = NULL,
                               round_number = NULL,
                               comp = "AFLM",
                               source = "AFL",
                               ...) {
  # Do some data checks
  # season <- check_season(season)
  check_comp_source(comp, source)

  dat <- switch(source,
    "AFL" = fetch_player_stats_afl(season, round_number, comp),
    "footywire" = fetch_player_stats_footywire(season, round_number, ...),
    "afltables" = fetch_player_stats_afltables(season, round_number),
    "fryzigg" = fetch_player_stats_fryzigg(season, round_number, comp),
    NULL
  )

  if (is.null(dat)) {
    cli::cli_warn("The source \"{source}\" does not have Player Stats.
                           Please use one of \"AFL\" \"footywire\", \"afltables\" or \"fryzigg\"")
  }
  return(dat)
}
```

jimmyday12 commented on R/fetch_team_stats.R:
```diff
@@ -0,0 +1,100 @@
+#' fetch_team_stats(2024)
+#' fetch_team_stats(2023, summary_type = "averages")
+fetch_team_stats <- function(season, summary_type = "totals") {
+  if (!is.numeric(season) || season < 1965 || season > 2025) {
```
You can follow the pattern we use in other functions. 

Example from `fetch_results`

```
  if (season < 1965) {
    cli::cli_abort("Season must be greater than 1965.You provided \"{season}\"")
  } 
```

Don't worry about data later than 2025, we don't want to have to update this each year with an explicit year. Just ensure that you have a good error message if you try access data from a year that doesn't exist on afltables

jimmyday12 commented on R/fetch_team_stats.R:
```diff
@@ -0,0 +1,100 @@
+  }
+  
+  team_name_map <- c(
+    "Adelaide" = "Adelaide",
```
You can remove this - we tend to just return data as is and let people join it if they like

jimmyday12 commented on R/fetch_team_stats.R:
```diff
@@ -0,0 +1,100 @@
+  url <- paste0("https://afltables.com/afl/stats/", season, "s.html")
+  page <- rvest::read_html(url)
+  tables <- page %>% rvest::html_elements("table")
+  if (length(tables) < 3) stop("Insufficient tables found on the page for season: ", season)
```
We've been using `cli_abort` for any stop messages

Example here that you can adapt

```
  if (season < 1965) {
    cli::cli_abort("Season must be greater than 1965.You provided \"{season}\"")
  }
```

jimmyday12 commented on R/fetch_team_stats.R:
```diff
@@ -0,0 +1,100 @@
+    "West Coast" = "West Coast",
+    "Footscray" = "Western Bulldogs"  # Standardize older data
+  )
+  
```
I've been using `cli_progress_step` at different points in functions to give nice status messages. 

For this one you could use something like

```
cli::cli_progress_step("Downloading Team stats for season {season} from AFLTables")
```

jimmyday12 reviewed (changes requested):
Thanks for the PR @DanielTomaro13 . This looks like it would be useful - have added a bunch of comments that are hopefully easy enough to follow. 

The main change is splitting this into two functions (`fetch_team_stats` and `fetch_team_stats_afltables`) to follow the conventions of other functions

DanielTomaro13 committed:
Add fetch_team_stats wrapper and afltables implementation with cli feedback, .data$ usage, and support for totals/averages

DanielTomaro13 commented:
Hi @jimmyday12 – thanks again for the detailed review and the step-by-step guidance! I really appreciate the support — I'm starting to explore development more seriously as it aligns with my career, and this has been a valuable learning experience.
I've now:
Refactored the function as requested into fetch_team_stats() (wrapper) and fetch_team_stats_afltables()
Updated all tidyselect expressions to use .data$
Replaced stop() with cli::cli_abort() and added cli::cli_progress_step() for clearer progress feedback
Removed the team_name_map from the final output and limited the use of a minimal name_map solely within the "averages" calculation to ensure the join with fetch_results_afltables() works correctly. The final output preserves the original AFLTables team names.
Let me know if there’s anything else you’d like adjusted — happy to keep refining!

jimmyday12 commented on R/fetch_team_stats.R:
```diff
@@ -0,0 +1,130 @@
+    dplyr::filter(.data$Team != "Totals") |>
+    dplyr::mutate(
+      dplyr::across(
+        where(is.character) & !dplyr::any_of(c("Team", "type")),
```
Just need to make sure you use `dplyr::where`

jimmyday12 commented on R/fetch_team_stats.R:
```diff
@@ -0,0 +1,130 @@
+      names()
+    
+    team_stats_final <- team_stats_final |>
+      dplyr::mutate(dplyr::across(all_of(numeric_cols), ~ . / .data$Games))
```
Just need to use `dplyr::all_of`

jimmyday12 commented on R/fetch_team_stats.R:
```diff
@@ -0,0 +1,130 @@
+    )
+    game_counts <- game_counts |>
+      dplyr::mutate(
+        Team = dplyr::recode(Team, !!!name_map)
```
Just need to use `.data$Team`

jimmyday12 reviewed (changes requested):
Thanks for all the changes @DanielTomaro13 - looks good. Made a few small requests. 

The only other thing is could you add a line item to `NEWS.md` and then run `devtools::check(document = TRUE)` to make sure the documentation updates and check if there are any errors or warnings

jimmyday12 reviewed (changes requested):
Looks good, made some suggestions for small changes. 

The only other thing is could you add a line item in `NEWS.md` with the changes and also run `devtools::check(document = TRUE)` so that documentation updates and you get a full check run

DanielTomaro13 committed:
small error changes and update NEWS.md

DanielTomaro13 commented:
Thanks!
Sorry about the small errors will look to prevent that in the future, just some habit stuff

DanielTomaro13 committed:
Update internal Rd for fetch_team_stats_afltables to fix codoc mismatch

DanielTomaro13 committed:
Regenerate Rd files for fetch_team_stats and fetch_team_stats_afltables to fix codoc mismatches

DanielTomaro13 committed:
Added fetch_team_stats to the list
